### ISSUE TITLE
[Feature Request] Parsing Target from Comment

### ISSUE THREAD

**Issue by NilsGruettner** [rvanbekkum/vsc-xliff-sync#37]
Hi,

first of all let me thank you for your awesome tool. It really helped streamlining the chore that is working with translation files.

**Description**

It would be great to have the option of parsing a translation from the comment, that you can specify with the label/caption in AL source file. Currently the comment is only added as "Note" to the .xlf.
This would enable the developer to specify a translation for each language when creating a label/caption and never think about it ever again. (like we are used to from C/AL :) )
The translation should only be parsed from the comment, if a new .xlf file is created with xliff sync or during syncing when new nodes are added. 

**Details**
* Add setup option "Parse target from comment"
* (Optional) Add a setup field to specify a separator (e.g. "|")
* The "Copy FromSource For Same Language" option should be overruled by this if a comment for the target-language is found, as we can expect an explicitely specified translation to have a higher priority than just the source text

**Expected result:**

`Caption = 'Customer No.',
            Comment = 'de-DE=Debitorennr.|en-US=Customer No.';`

will look in .xlf like this:

```
...
<file datatype="xml" source-language="en-US" target-language="en-US">
...
<trans-unit id="Table 698291618 - Field 4190130733 - Property 2879900210" size-unit="char" translate="yes" xml:space="preserve">
          <source>Customer No.</source>
          <target state="translated">Customer No.</target>
          <note from="Developer" annotates="general" priority="2">de-DE=Debitorennr.|en-US=Customer No.</note>
        </trans-unit>
```

```
...
<file datatype="xml" source-language="en-US" target-language="de-DE >
...
<trans-unit id="Table 698291618 - Field 4190130733 - Property 2879900210" size-unit="char" translate="yes" xml:space="preserve">
          <source>Customer No.</source>
          <target state="translated">Debitorennr.</target>
          <note from="Developer" annotates="general" priority="2">de-DE=Debitorennr.|en-US=Customer No.</note>
        </trans-unit>
```

**Comment by rvanbekkum** [comment]
Hi @NilsGruettner,
Thank you for your feature request and for the clear description. It's also nice to hear you find the extension useful :).

I'm curious to hear what problems you are running into that could be solved with the addition of this new feature?
If I understand correctly, you will have translations in the source code again (just like with the `CaptionML` property in AL) and also corresponding XLIFF files with (initial) translations coming from the notes/comments in the source code.
I can think of some reasons you'd rather use `Caption` instead of `CaptionML`, e.g., being able to utilize various XLIFF tooling, still allowing translations for new languages to be added without changing the AL source code.
A downside of adding translations in the `Comment` however is that you will have a change in your AL source code (`.al` files) anytime you want to change the translation in there.

Could you give me an idea of what the benefits of having (initial) translations in the AL source code as well could be?
A benefit I could think of is that you can directly see how a field/action/group is translated and don't have to match it with the trans-unit and enter the translation later on, although I can also think of other solutions for that problem.

Hope to see your reply!

**Comment by NilsGruettner** [comment]
Hi @rvanbekkum,

thank you for your reply.

I understand your point, that having translations in source code is definitely not best practice.
But having an initial translation in source code also brings multiple advantages and in some cases is not too bad:

* Fiddling with the "Multilingual Editor" provided by Microsoft is a pain. I know there are other tools that can help, but the process is still annoying
* Having to deal with translations in an external tool adds another step to development process. Being able to provide an initial translation in source code gets rid of this step and makes development faster
* The developer, who is usually proficient in business terms used in BC, most of the time can provide the best translation (at least for English and native language). It is convenient to provide that in source code, using code snippets
* Currently, I add translations in bulk towards the end of development of an extension to reduce overhead. With initial translations from source code I can provide a fully translated extension to software testers with each code change
* (as you mentioned) you could still add translations for new languages without changing source code
* You still have the possibility to submit .xlf files to external translation agencies (e.g. for translating into a new language or QA)

And after all, the user does not really care. Experience has shown, that users do not expect a flawless translation, as long as they understand what is going on on screen. Considering the already high quality of translations usually provided by a proficient developer, an initial translation will likely never be touched again, so having it in source code will not neccessarily lead to continous code changes.

**Comment by rvanbekkum** [comment]
Thank you for your clear explanation. I will implement the suggested feature for it to become available in a next version.

**Comment by NilsGruettner** [comment]
That's awesome! Thank you a lot.

**Comment by rvanbekkum** [comment]
Hi @NilsGruettner,
I have implemented the feature and hope you could have a look.
I have added two new settings `"xliffSync.parseFromDeveloperNote"` (default: false) and `"xliffSync.parseFromDeveloperNoteSeparator"` (default: "|").
The changes have been pushed to a commit in the [`feature/37-parse-target-from-note`](https://github.com/rvanbekkum/vsc-xliff-sync/tree/feature/37-parse-target-from-note) branch.

**Here is an example:**
I have added the following settings to the `settings.json` file of my project:
```json
{
    "xliffSync.copyFromSourceForSameLanguage": true,
    "xliffSync.parseFromDeveloperNote": true,
    "xliffSync.parseFromDeveloperNoteSeparator": "|"
}
```
Please note that "|" is also the default separator.

Next, I have set up the following sample source code:
```pascal
procedure MyProcedure()
var
    DissapointDaveErr: Label 'I am afraid I can not do that, Dave', Comment = 'en-US=I am trepidacious I can not do that, Dave|nl-NL=Ik ben bang dat ik dat niet kan doen, Dave';
    ExtremeMathMsg: Label 'One plus one = two', Comment = 'en-US=One plus one = TWO|nl-NL=Een plus een is twee|nl-BE=Een plus een is twee';
    SatisfySamMsg: Label 'I hope you are satisfied, Sam', Comment = 'This is just a regular comment';
    TerminateTimErr: Label 'Hasta la vista, Tim', Comment = 'nl-BE=Salut, Tim';
    WorshipWillemMsg: Label 'Thank you Willem';
begin
    Error(DissapointDaveErr);
end;
```
This sample contains labels for the following:
1. A label with a en-US and nl-NL translation in the comment. (After sync., the en-US file should take the translation from the Developer note and not from the source)
2. A label with en-US, nl-BE, nl-NL translations with a '=' in the translations. (The "=" should be included in each translation)
3. A label with a regular comment (without translations).
4. A label with a single nl-BE translation
5. A label without any comment


After running "AL: Package" (build/compile) to get the `.g.xlf` generated base-XLIFF file, and running a sync for our target languages, we get the following files:

en-US:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="urn:oasis:names:tc:xliff:document:1.2 xliff-core-1.2-transitional.xsd">
  <file datatype="xml" source-language="en-US" target-language="en-US" original="ALProject7">
    <body>
      <group id="body">
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 1500874637" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>I am afraid I can not do that, Dave</source>
          <target state="translated">I am trepidacious I can not do that, Dave</target>
          <note from="Developer" annotates="general" priority="2">en-US=I am trepidacious I can not do that, Dave|nl-NL=Ik ben bang dat ik dat niet kan doen, Dave</note>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType DissapointDaveErr</note>
        </trans-unit>
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 1441641737" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>One plus one = two</source>
          <target state="translated">One plus one = TWO</target>
          <note from="Developer" annotates="general" priority="2">en-US=One plus one = TWO|nl-NL=Een plus een is twee|nl-BE=Een plus een is twee</note>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType ExtremeMathMsg</note>
        </trans-unit>
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 1663451319" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>I hope you are satisfied, Sam</source>
          <target state="translated">I hope you are satisfied, Sam</target>
          <note from="Developer" annotates="general" priority="2">This is just a regular comment</note>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType SatisfySamMsg</note>
        </trans-unit>
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 3882606824" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>Hasta la vista, Tim</source>
          <target state="translated">Hasta la vista, Tim</target>
          <note from="Developer" annotates="general" priority="2">nl-BE=Salut, Tim</note>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType TerminateTimErr</note>
        </trans-unit>
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 4184091849" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>Thank you Willem</source>
          <target state="translated">Thank you Willem</target>
          <note from="Developer" annotates="general" priority="2"/>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType WorshipWillemMsg</note>
        </trans-unit>
      </group>
    </body>
  </file>
</xliff>
```

nl-BE:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="urn:oasis:names:tc:xliff:document:1.2 xliff-core-1.2-transitional.xsd">
  <file datatype="xml" source-language="en-US" target-language="nl-BE" original="ALProject7">
    <body>
      <group id="body">
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 1500874637" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>I am afraid I can not do that, Dave</source>
          <target state="needs-translation"/>
          <note from="Developer" annotates="general" priority="2">en-US=I am trepidacious I can not do that, Dave|nl-NL=Ik ben bang dat ik dat niet kan doen, Dave</note>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType DissapointDaveErr</note>
        </trans-unit>
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 1441641737" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>One plus one = two</source>
          <target state="translated">Een plus een is twee</target>
          <note from="Developer" annotates="general" priority="2">en-US=One plus one = TWO|nl-NL=Een plus een is twee|nl-BE=Een plus een is twee</note>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType ExtremeMathMsg</note>
        </trans-unit>
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 1663451319" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>I hope you are satisfied, Sam</source>
          <target state="needs-translation"/>
          <note from="Developer" annotates="general" priority="2">This is just a regular comment</note>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType SatisfySamMsg</note>
        </trans-unit>
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 3882606824" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>Hasta la vista, Tim</source>
          <target state="translated">Salut, Tim</target>
          <note from="Developer" annotates="general" priority="2">nl-BE=Salut, Tim</note>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType TerminateTimErr</note>
        </trans-unit>
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 4184091849" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>Thank you Willem</source>
          <target state="needs-translation"/>
          <note from="Developer" annotates="general" priority="2"/>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType WorshipWillemMsg</note>
        </trans-unit>
      </group>
    </body>
  </file>
</xliff>
```

nl-NL:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="urn:oasis:names:tc:xliff:document:1.2 xliff-core-1.2-transitional.xsd">
  <file datatype="xml" source-language="en-US" target-language="nl-NL" original="ALProject7">
    <body>
      <group id="body">
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 1500874637" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>I am afraid I can not do that, Dave</source>
          <target state="translated">Ik ben bang dat ik dat niet kan doen, Dave</target>
          <note from="Developer" annotates="general" priority="2">en-US=I am trepidacious I can not do that, Dave|nl-NL=Ik ben bang dat ik dat niet kan doen, Dave</note>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType DissapointDaveErr</note>
        </trans-unit>
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 1441641737" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>One plus one = two</source>
          <target state="translated">Een plus een is twee</target>
          <note from="Developer" annotates="general" priority="2">en-US=One plus one = TWO|nl-NL=Een plus een is twee|nl-BE=Een plus een is twee</note>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType ExtremeMathMsg</note>
        </trans-unit>
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 1663451319" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>I hope you are satisfied, Sam</source>
          <target state="needs-translation"/>
          <note from="Developer" annotates="general" priority="2">This is just a regular comment</note>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType SatisfySamMsg</note>
        </trans-unit>
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 3882606824" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>Hasta la vista, Tim</source>
          <target state="needs-translation"/>
          <note from="Developer" annotates="general" priority="2">nl-BE=Salut, Tim</note>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType TerminateTimErr</note>
        </trans-unit>
        <trans-unit id="PageExtension 1255613137 - Method 3998599243 - NamedType 4184091849" size-unit="char" translate="yes" xml:space="preserve" al-object-target="Page 2901867346">
          <source>Thank you Willem</source>
          <target state="needs-translation"/>
          <note from="Developer" annotates="general" priority="2"/>
          <note from="Xliff Generator" annotates="general" priority="3">PageExtension CustomerListExt - Method MyProcedure - NamedType WorshipWillemMsg</note>
        </trans-unit>
      </group>
    </body>
  </file>
</xliff>
```

Could you please review whether this matches your request, or if I forgot something? After that, I will merge the changes into the master branch and will create a release.
Thanks!

**Comment by NilsGruettner** [comment]
Hi @rvanbekkum,

your example seems perfect. It exactly matches my needs.
Thank you a lot for your effort.
I am really looking forward to try this out!

Also thanks for providing a good giggle with those brilliant text samples :)

**Comment by rvanbekkum** [comment]
This change is shipped in version 0.3.5.
